### ISSUE TITLE
Re-balance swarm tasks when new nodes join

### ISSUE THREAD

**Issue by F21** [docker/docker#24103]
It would be nice to be able to automatically re-balance tasks when a new node joins.

For example, we have 4 nodes, running a redis service with 5 tasks. When a new node joins, swarm should automatically re-balance the tasks so that all nodes will be running 1 redis task.

**Comment by stevvooe** [comment]
This is something we've discussed but has been omitted from 1.12. There are problems with new nodes getting "mugged" as they are added. We also avoid pre-emption of healthy tasks. Rebalancing is done over time, rather than killing working processes. Pre-emption is being considered for the future.

As a workaround, scaling a service up and down should rebalance the tasks. You can also trigger a rolling update, as that will reschedule new tasks.

cc @aluzzardi @aaronlehmann

**Comment by aaronlehmann** [comment]
See https://github.com/docker/swarmkit/issues/310 for the relevant swarmkit issue.

**Comment by adrianovieira** [comment]
Yes! This is a feature I'd like see on docker swarm as I've filed on #24417

**Comment by rogaha** [comment]
+1, meanwhile you can scale down and up to get them re-scheduled properly.

**Comment by omerh** [comment]
I'm seeing this feature helpful not when a node comes back to life, but more when scaling up a swarm cluster.

I am doing autoscale in AWS and the scale up is based on CPU usage
So when autoscale provision a new node it will be idle.
I think that the possibility to control re-balancing when running the join command would be great.

```
docker swarm join ... --rebalance 
```     

I would also suggest soft / hard re-balance.
Soft -  Aligning the spread of containers across the swarm cluster
Hard - According to resources, which will probably will be harder to implement

**Comment by srigi** [comment]
I don't think that automatic rebalancing is good. I want to have a some control about it. The solution suggested in #28989 (force rebalancing, using `docker service update --force <service>` in 1.13) works fine.

**Comment by libby2cm** [comment]
So, what's the status/thoughts on this?  REALLY need some sort of mechanism to kick this off manually as needed.  At minimum, a 'docker node rebalance' command or something to do it ad-hoc after rebooting hosts would be great.  Longer term, a feature akin to VMW DRS with adjustable threshholds would be nice.

**Comment by markvr** [comment]
The problem with doing `docker service update --force <service>` is you obviously need to do it for every service on the swarm individually.  A `docker swarm rebalance` command would be really helpful.

**Comment by cpuguy83** [comment]
@markvr But I still think in this case you'd want to do it explicitly for each service like... `docker swarm rebalance serviceA serviceB serviceC` rather than on all services.

**Comment by srigi** [comment]
Better place for this command would be `docker stack`. IMHO.

**Comment by markvr** [comment]
@cpuguy83 
I guess there are different use cases.  I'm happy for Docker to place services where it wants, based on the constraints I have defined.  Having to list all the services would be possible but I would have to write a python script to call the API, obtain a list of services and construct the command with all the services in and run that.  Having that functionality built into the CLI directly would make things easier.

@srigi having it at a stack level would be better than individual services, but still means you need to iterate a list of all stacks and construct a command if you want to enable Docker to rebalance the whole swarm.

Could of course have it on all options!

  - `docker service rebalance`
  - `docker stack rebalance` 
  - `docker swarm rebalance`

**Comment by libby2cm** [comment]
For anyone interested, I am handling it via this one liner .. useful after node maintenance operations:

`docker service ls -q > dkr_svcs && for i in `cat dkr_svcs`; do docker service update "$i" --detach=false --force ; done`

**Comment by msfidelis** [comment]
This feature makes it possible to use Swarm in production using Load Balance and Auto Scaling in AWS. 
I can not safely deploy my apps for this reason

![Docker](http://i.imgur.com/DQh6jQm.png)

**Comment by thaJeztah** [comment]
@msfidelis what version of docker are you running? Also, if you want a service to be deployed on each node, using `--mode=global` may be what you're looking for

**Comment by msfidelis** [comment]
@thaJeztah I'm using Docker in version `Docker version 17.03.1-ce, build 7392c3b/17.03.1-ce`. 

The `--mode=global` option does not allow using the `docker service scale servicename=10` or` docker service update --replicas = 50 servicename`

**Comment by thaJeztah** [comment]
@msfidelis I understand; however, you're scaling nodes. Setting `mode=global` would automatically "scale" your service based on the number of nodes. If you need extra capacity, scale the nodes, and get more instances of your service, scale down nodes, and you get less instances of your service.

**Comment by msfidelis** [comment]
@thaJeztah I believe that this is not a solution to the problem. We need to up 10 containers per instance. Scale 1 instance per container is a very expensive solution.

**Comment by thaJeztah** [comment]
Trying to understand your use case;

Your example shows (visible) 6 instances on two nodes. How would rebalancing those 6 instances over five nodes resolve that problem? If more capacity is needed, both more nodes need to be added, and the service need to be scaled so that more instances are deployed. Scaling up will deploy those instances on the new nodes. You may be able to hook this into your auto-scaling (i.e., when scale up nodes, also scale up the service)

If autoscaling is configured based on "load" per node, and docker would only _rebalance_ the existing instances; the load per node would drop (assuming CPU/Memory if services is restricted), potentially leading to autoscale scaling down.

**Comment by markvr** [comment]
I've been trying to find some info on how Docker schedules tasks, and there isn't a huge amount around, the nearest I found was https://github.com/docker/swarmkit/issues/308

I did some tests with 2 nodes with 2Gb RAM and 3 services with memory reservations:
A = 750Mb
B = 750Mb
C = 1500Mb

If I create A & B first they are scheduled onto each node as expected, but then C won't run because it doesn't have a node with enough free RAM, even though if the containers were reorganised, it would fit.  Running `docker service update --force` didn't make any difference to the distribution.

If I create C first, then A & B everything fits and runs OK.   This makes it a difficult manual process to calculate the "binpacking" of all the services, and what order to start them in to maximise the host resource us, or how to use constraints to "pin" them to each node.  Ironically this was one of the key things I was hoping to get from Docker (and hadn't realised it lacked), as our current non-Docker platform requires all this work!

If there was a `docker swarm rebalance` command that could do this automatically do this it would be really great, and even better would be if Docker calculated this automatically when services were created to maximise the chances of the new service being run.

There appears to be a "binpack" scheduler on the old swarm, but I can't find any docs on whether this can be used still: https://docs.docker.com/swarm/scheduler/strategy/#binpack-strategy-example

**Comment by thaJeztah** [comment]
I can see the use-case of that, but also a danger, i.e., deploying service `C` would result in service `A` and/or `B` temporarily becoming unavailable (depending on the situation of course). Should a new service be able to affect existing services?

I think that _if_ this is implemented, it should be an explicit action, not something to do without a user explicitly requesting this (per my comments above)

> There appears to be a "binpack" scheduler on the old swarm

I don't think the old swarm did rescheduling / shuffling running services to make them fit, but I may be wrong there.

**Comment by markvr** [comment]
Totally agree that the default behaviour should remain as currently.  The way I could see this working, is in my situation above where C is pending, then a new command `docker swarm rebalance` would calculate the most efficient packing, and then reschedule containers as needed. 

This could presumably have modifiers like `--start-first` , using healthchecks to do zero downtime rescheduling, assuming there is enough headroom in the cluster, or `--dry-run` to report on what it would do.

Additionally, you could have `docker swarm create --rebalance ...` to automatically rebalance the cluster when creating a new service.

I realise this is a very non-trivial amount of work!  As well as calculating the end binpacking configuration, you would need to calculate how to actually get there whilst respecting constraints, and maintaining the spread of replicas within a service across nodes/fault domains etc like the existing HA scheduler does.  I don't know how I would even begin to do that!

Alternatively, are there plans to have scheduler plugins?  It feels like everyone has their own favourite way to schedule services,, and that would get this all out of your hair!

I guess you could then do: `docker swarm rebalance --scheduler myfavouritealgorithm` ?

**Comment by thaJeztah** [comment]
I think there's an issue about having scheduler plugins in SwarmKit (replying from my phone, so can't quickly search right now)

**Comment by segler-alex** [comment]
i have another proposal: a combination of global mode with scaling applied.
this would mean, that in your docker-compose file you could do something like:

web:
   image: nginx
   deploy:
      replicas: 3
      mode: global

this would automatically ensure 3 running containers on each added host.
so the total number of containers will grow if the number of machines grows.
and it will decrease if the total number of machines decrease.
this would greatly help on doing some kind of scaling a service.

**Comment by eyz** [comment]
Without looking at the underlying code, it seems like a rebalance would be:
1) generate the broad plan for service placement (using the existing code for scheduling task placement)
2) compare current task placement for the service with the service placement plan from step 1 (not enforcing certain slot IDs onto certain nodes, but rather only considering the presence of healthy tasks on the nodes against the broad placement plan)
3) where current tasks should not have the desired count of service tasks (per the plan), then schedule stops in the rebalance operation queue
4) where current tasks should have services, but do not have the proper count, then schedule starts in the rebalance operation queue

Does this capture the general concept?

**Comment by getvivekv** [comment]
I have had this same requirement, where I have a custom AWS autoscaling script developed, it would automatically provision new ec2 instances however the docker swarm wouldn't auto rebalance the services causing the recently joined node to be idle. 

I am in favor of implementing a built-in feature in Docker Swarm to auto rebalance the running services, provided that the user know what exactly the feature is and also the problems it can create for certain jobs/containers and make a decision based on their own use case. 

For example, if there is a container that runs a long-running process(started by a cronjob) then auto-rebalancing will kill that running container when rescheduling it on another node causing the process to be terminated abruptly and the process may be executed only next time the cronjob runs, which may be next day or next week. 

On the other hand, if the containers are for processes like serving a static page then auto-rebalancing wouldn't affect anything and there are several users for this use case and I believe this feature would be useful for them. 

I've come up with this workaround on this. Here my use case is static website running from containers of several replicas and setup as AWS AutoScaling. The following script now helps me rebalance the containers automatically when new node joins the cluster.

This script is installed on the manager node and added as a cronjob that runs every 5 minutes

```
FILE=/tmp/worker.nodes
for node in `docker node ls --filter role=worker -q`; do
    if grep -Fxq "${node}" ${FILE}
    then
        echo "This node ${node} already exists"
    else
        echo "This node ${node} joined recently, so rebalance"
        for service in `docker service ls -q`; do
          docker service update --with-registry-auth  --detach=true --force ${service}
        done
    fi
done

docker node ls --filter role=worker -q > ${FILE}
```

This is kind of a dirty hack, but it does the job for me. It uses the `docker service update --force` to rebalance and also save the current ids of the worker nodes. This can further be adjusted to rebalance a specific docker service by name if required (as I have )

It would be great if this feature can be built in, in the service meta data so that we can define which services we would like to rebalance when a new node joins the cluster.

**Comment by ericrange** [comment]
I dont understand this discussion. If a node comes back there should!!! be an auto rebalancing. Otherwise i wouldnt call it "docker orchestration". if you ran it and then stop and never runs again on a node which was down: there is already something for this, called nohup. why do i need docker swarm for this?
=> serious question. im a bit disappointed that docker swarm can not handle this :(
how can i run production apps with this? when i can not be sure if the stack runs anyway. i mean, there could be some network problems for a moment... and then? complete stack down? nice...
im not sure if this is really the case. (im currently deploying docker swarm clusters => newbe) but what i have heard and read so far, makes me very nervous!! :(

**Comment by ericrange** [comment]
and i also dont understand why rebalancing causes problems? if a nodes come back, lets say node "2":
1. launch container there (node 2)
2. redirect traffic from from container 2 on node 1 to container 1 on node 2
3. shutdown container 2 on node 1

**Comment by cpuguy83** [comment]
@ericrange The feature just wasn't added. The "discussion" is "we decided not to include this in the initial release", and no one has picked up the work since. This issue is open, as is the referenced issue on swarmkit.

**Comment by MGD1981** [comment]
Vivek: I have a quick and dirty script to rebalance as well, without
needing the iteration. For one service, which lets you run it using the
service name as an argument:

#!/bin/bash
sudo /bin/docker service update $1 --force --detach=false

For all services:
docker service ls -q | xargs -i docker service update {} --force
--detach=false

A word of warning though: at first I was detaching like you are, but I
realized that my swarm was having a lot of trouble handling all those
updates at once, so now it doesn't detach. Much slower, but safer.

**Comment by msfidelis** [comment]
I create a similar script on master's crontab. But, this is don't scale. 

We migrate our applications to Kubernetes Cluster to work with AWS ASG. It's working. 

I hope a similar solution for Swarm in the future. Until now, this is don't work for a large applications scaling between performance and costs on Cloud. 😢

**Comment by jay915** [comment]
This is why you should use Kubernetes on production

**Comment by cpuguy83** [comment]
@jay915 To my knowledge, Kubernetes does not rebalance pods. You have to either update the pod spec so that pods are re-scheduled, or kill the pods on nodes you want rebalanced.

**Comment by cpuguy83** [comment]
Also meant add, please point me in to articles or docs if I'm mistaken on that, would love to see! 👼

**Comment by govorusha** [comment]
This is a game breaker for big clusters. When CPU/MEM limits defined for each service, swarm should orchestrate a number of services per node automatically.

**Comment by cpuguy83** [comment]
@govorusha that's exactly what swarm does, it just doesn't automatically rebalance when new nodes come online, because it prefers to keep things running rather than kill things to move them to a new node.

This is the same behavior as k8s unless I'm missing something.

**Comment by JohnTheNerd** [comment]
I feel like the perfect resolution for this would be to re-balance containers that have their update order set as start-first automatically, as they will suffer no downtime regardless. wouldn't this be a win-win for everyone?

**Comment by thaJeztah** [comment]
> I feel like the perfect resolution for this would be to re-balance containers that have their update order set as start-first automatically, as they will suffer no downtime regardless. wouldn't this be a win-win for everyone?

What would the benefit be of starting new instances on the new node, and killing the existing, _healthy_ instances?

There's also scenarios where this could cause other disruptions, for example, I could add new nodes to a cluster in preparation of deploying a new service. With the proposed change, swarm would start deploying _existing_ services on the new node (and killing tasks on other nodes) the moment the new node joins, potentially leading a "race condition" with the new service I'm about to deploy.

**Comment by magnussuther** [comment]
The scenario of adding a new node to the cluster prior to deploying new services certainly is a valid one. However, this leads to a segmented cluster, where most or all of 'Service B' exists in the AZ where the new node lives. Might not be an issue for some...
Anyway, to prevent old services from jumping to your new node you should be able to use Placement Constraints.


> What would the benefit be of starting new instances on the new node, and killing the existing, healthy instances?

One benefit would be to reduce load on overloaded nodes automatically, such that we can use AWS Autoscaling for example. This is the problem I've been trying to bend my head around for a couple of days.

Scenario: 

One or 2 of my 2 running Worker nodes are running high on CPU usage (over 20% avg for a couple of hours), and I risk running out of CPU Credits. So my auto-scaler boots a new instance, and the Launch Template runs my user-data script which joins the new node to the Swarm as a Worker. So far, all is perfectly fine.

Obviously, launching a new instance and not having any containers assigned to it is rather useless and a waste of electrons.

The thing missing in Docker Swarm is the ability to reduce load on the 2 overloaded nodes in a glitch-free and automatic manner. Basically, the only way my 3:rd node gets assigned any containers automatically is if a container dies on any of the other nodes and that doesn't happen very often.
In addition, there doesn't seem to be any way to instruct Docker Swarm to consider a node as overloaded by defining thresholds for various types of loads. For example, I would like to tell Swarm to perform rebalancing if the CPU load is 20% avg for a hour or so, or if RAM is 85% avg for 15 mins)

I am running `docker stack deploy` every minute in the form of a crontab, so in my case there are 2 additional scenarios that may assign containers to my 3:rd node:
- a service definition is updated in the docker-compose.yml in a git repo
- a new Docker Image matching the defined tag is available
... and that doesn't happen very often either.


So, after reading the comments above, it seems like the following approach is needed.
- Register a crontab on each of the Managers, that
  - checks if the current node is the Leader, else abort
  - checks if new Workers have been added since the last run, else abort
  - executes `docker service ls -q | xargs -i docker service update {} --force --detach=false`
  
From what I understand, there's no way to prevent `docker service update --force` from restarting _all_ containers for that service, causing unwanted disturbances.

Right now there's too much blood in my caffeine veins... but in my mind, this is the way things should work:
- There should be a setting for `min_replicas` for each service.
- Rebalancing should only happen for services with more than `min_replicas` running, unless rebalancing is forced. `min_replicas` may be set to 1, but a user may want to increase that number if the service chokes due to load during the time it takes to rebalance, or if the user wants to reduce the chance that the rebalancing process touches a particular service. The service is left untouched if `min_replicas == replicas`.
- Swarm should select `(replicas - min_replicas)` number of containers, and
  - Drain traffic to those containers
  - Move all or some of those containers, but avoid murdering the new node with tasks
  - Start routing traffic again


I do realize that this is not a trivial thing to build, and there are other tools that may experience a moment of confusion when containers are jumping around. Since I'm using Traefik, I'm worried that it may take a few seconds for Traefik to understand what just happened, and if there's a glitch here then we're back to square one.

**Comment by johackim** [comment]
Thanks @libby2cm !

Here an improve : 

```bash
for service in $(docker service ls -q); do docker service update --force $service; done
```
